classes/doi/Repository.inc.php

```diff
@@ -0,0 +1,440 @@
+use PKP\context\Context;
+use PKP\plugins\HookRegistry;
+use PKP\plugins\Plugin;
+use PKP\plugins\PluginRegistry;
+use PKP\services\PKPSchemaService;
+use PKP\submission\PKPSubmission;
+use PKP\validation\ValidatorFactory;
+
+class Repository
+{
+    public const SETTING_ENABLE_DOIS = 'enableDois';
+    public const SETTING_ENABLED_DOI_TYPES = 'enabledDoiTypes';
+    public const SETTING_DOI_PREFIX = 'doiPrefix';
+    public const SETTING_USE_DEFAULT_DOI_SUFFIX = 'useDefaultDoiSuffix';
+    public const SETTING_CUSTOM_DOI_SUFFIX_TYPE = 'customDoiSuffixType';
+    public const SETTING_CONFIGURED_REGISTRATION_AGENCY = 'registrationAgency';
+    public const SETTING_NO_REGISTRATION_AGENCY = 'none';
+
+    public const CUSTOM_SUFFIX_LEGACY = 'legacy';
+    public const CUSTOM_SUFFIX_LEGACY_USER = 'legacyCustom';
+    public const CUSTOM_SUFFIX_MANUAL = 'customId';
+
+    public const LEGACY_CUSTOM_PUBLICATION_PATTERN = 'doiPublicationSuffixPattern';
+    public const LEGACY_CUSTOM_REPRESENTATION_PATTERN = 'doiRepresentationSuffixPattern';
+
+    /** @var DAO $dao */
+    public $dao;
+
+    /** @var string $schemaMap The name of the class to map this entity to its schema */
+    public $schemaMap = maps\Schema::class;
+
+    /** @var Request $request */
+    protected $request;
+
+    /** @var PKPSchemaService $schemaService */
+    protected $schemaService;
+
+    private ?Plugin $_registrationAgencyPlugin = null;
+
+    public function __construct(DAO $dao, Request $request, PKPSchemaService $schemaService)
+    {
+        $this->dao = $dao;
+        $this->request = $request;
+        $this->schemaService = $schemaService;
+    }
+
+    /** @copydoc DAO::newDataObject() */
+    public function newDataObject(array $params = []): Doi
+    {
+        $doi = $this->dao->newDataObject();
+        if (!empty($params)) {
+            $doi->setAllData($params);
+        }
+        return $doi;
+    }
+
+    /** @copydoc::get() */
+    public function get(int $id): ?Doi
+    {
+        return $this->dao->get($id);
+    }
+
+    /** @copydoc DAO::getCount() */
+    public function getCount(Collector $query): int
+    {
+        return $this->dao->getCount($query);
+    }
+
+    /** @copydoc DAO::getIds() */
+    public function getIds(Collector $query): Collection
+    {
+        return $this->dao->getIds($query);
+    }
+
+    /** @copydoc DAO::getMany() */
+    public function getMany(Collector $query): LazyCollection
+    {
+        return $this->dao->getMany($query);
+    }
+
+    /** @copydoc DAO::getCollector */
+    public function getCollector(): Collector
+    {
+        return App::make(Collector::class);
+    }
+
+    /**
+     * Get an instance of the map class for mapping
+     * DOIs to their schema
+     */
+    public function getSchemaMap(): maps\Schema
+    {
+        return app('maps')->withExtensions($this->schemaMap);
+    }
+
+    /**
+     * Check if duplicate of this DOI has already been recorded across all contexts.
```

Comment:
This doesn't really change the question but add some further context: this in particular was thinking about Datacite's Cool DOIs pattern in an instance where a single prefix was used for multiple journals. If the journal slug/path were no longer used in the DOI suffix, e.g. 10.9876/ABCD-1234, it would be important to catch that a duplicate suffix weren't generated and already in use. Would it be better to check across all contexts for cases like these while limiting the check to the specific context for others? I suppose what's actually important in this case is not checking across contexts but across all uses of a single prefix.

Reply:
Hmm, I think we shouldn't try to take on responsibility for this. A service provider operating the same prefix across multiple journals could handle this by entering their prefix with an addition to their "official" prefix like this: `10.1234/<my-unique-journal>-`.

In general, we are moving towards keeping journals completely distinct from each other, so I'd prefer to force the service provider to handle this in their configuration.